File: apps/windows/TeslaSync.App/feature-views/WebhookChannelsSection/WebhookChannelsSection.Source.cs
using System.Collections.Generic;
using System.Globalization;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Text.Json.Nodes;
using TeslaSync.App.Core.Data;
using TeslaSync.App.Core.Data.Behavior;
using TeslaSync.App.Core.Data.Net;
using TeslaSync.App.Core.Data.State;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The data port the <see cref="WebhookChannelsViewModel"/> binds to (P1/S8 state-holder seam). It yields the
/// cache-then-network channel read the web <c>useWebhookChannels()</c> composes (the kind=webhook slice of
/// <c>GET /notifications</c>) and exposes the five mutations/utilities the section invokes: save, delete, toggle,
/// the HMAC-aware webhook test (web <c>useTestWebhookChannel</c>) and the signature preview utility (web
/// <c>useWebhookSignaturePreview</c>). The view never performs HTTP itself; the concrete
/// <see cref="WebhookChannelsSource"/> (or a test fake) drives this.
/// </summary>
public interface IWebhookChannelsSource
{
    /// <summary>Stream the cache-then-network webhook-list snapshots (kind=webhook filtered), newest cache first.</summary>
    IAsyncEnumerable<RepositoryResult<WebhookChannelList>> StreamWebhooksAsync(CancellationToken cancellationToken = default);

    /// <summary>Create (<c>POST</c>) or update (<c>PUT</c>) a webhook from <paramref name="body"/> (web <c>useSaveChannel</c>).</summary>
    Task SaveAsync(JsonObject body, long? id, CancellationToken cancellationToken = default);

    /// <summary>Delete a webhook (web <c>useDeleteChannel</c>).</summary>
    Task DeleteAsync(long id, CancellationToken cancellationToken = default);

    /// <summary>Flip a webhook's enabled flag (web <c>useToggleChannel</c>).</summary>
    Task ToggleAsync(long id, CancellationToken cancellationToken = default);

    /// <summary>Fire a structured test delivery through the HMAC-aware webhook path (web <c>useTestWebhookChannel</c>).</summary>
    Task<WebhookTestResult> TestWebhookAsync(long id, CancellationToken cancellationToken = default);

    /// <summary>Ask the server to compute the X-TeslaSync-Signature for (secret, body) (web <c>useWebhookSignaturePreview</c>).</summary>
    Task<string> PreviewSignatureAsync(string secret, string body, CancellationToken cancellationToken = default);
}

/// <summary>
/// The repository-backed <see cref="IWebhookChannelsSource"/> — the native data adapter for the WebhookChannels
/// surface. The read replays cache-then-network through the shared <see cref="CacheThenNetworkEngine"/> (the
/// channel list is cached as raw JSON under the same <c>notifications:channels</c> key the generic channels
/// surface uses, then mapped + kind=webhook-filtered by <see cref="WebhookChannelsResultMapper"/>); the four
/// mutations and the signature-preview utility go straight through the generated contract client. No HTTP
/// touches the view.
/// </summary>
public sealed class WebhookChannelsSource : IWebhookChannelsSource
{
    private const string ChannelsCacheKey = "notifications:channels";

    private static readonly ApiRequest ChannelsRequest = new(WebhookChannelsRegistration.ChannelsOperation);

    private readonly IApiClient _api;
    private readonly CacheThenNetworkEngine _engine;
    private readonly JsonSerializerOptions _json;

    /// <summary>Creates the source over the contract client, cache-then-network engine and JSON settings.</summary>
    public WebhookChannelsSource(IApiClient api, CacheThenNetworkEngine engine, ApiClientOptions options)
    {
        ArgumentNullException.ThrowIfNull(api);
        ArgumentNullException.ThrowIfNull(engine);
        ArgumentNullException.ThrowIfNull(options);
        _api = api;
        _engine = engine;
        _json = options.Json;
    }

    /// <inheritdoc />
    public async IAsyncEnumerable<RepositoryResult<WebhookChannelList>> StreamWebhooksAsync(
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        var raw = _engine.StreamAsync<JsonElement>(
            ChannelsCacheKey,
            ct => _api.SendAsync<JsonElement>(ChannelsRequest, ct),
            IsEmptyWebhooks,
            _json,
            CacheFreshness.LiveStaleSeconds,
            cancellationToken);

        await foreach (var emission in raw.ConfigureAwait(false))
        {
            yield return WebhookChannelsResultMapper.MapWebhooks(emission);
        }
    }

    /// <inheritdoc />
    public async Task SaveAsync(JsonObject body, long? id, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(body);

        var request = id is { } channelId
            ? new ApiRequest(WebhookChannelsRegistration.UpdateOperation, PathParams: PathFor(channelId), Body: body)
            : new ApiRequest(WebhookChannelsRegistration.CreateOperation, Body: body);

        await _api.SendAsync<JsonElement?>(request, cancellationToken).ConfigureAwait(false);
    }

    /// <inheritdoc />
    public async Task DeleteAsync(long id, CancellationToken cancellationToken = default)
    {
        var request = new ApiRequest(WebhookChannelsRegistration.DeleteOperation, PathParams: PathFor(id));
        await _api.SendAsync<JsonElement?>(request, cancellationToken).ConfigureAwait(false);
    }

    /// <inheritdoc />
    public async Task ToggleAsync(long id, CancellationToken cancellationToken = default)
    {
        var request = new ApiRequest(WebhookChannelsRegistration.ToggleOperation, PathParams: PathFor(id));
        await _api.SendAsync<JsonElement?>(request, cancellationToken).ConfigureAwait(false);
    }

    /// <inheritdoc />
    public async Task<WebhookTestResult> TestWebhookAsync(long id, CancellationToken cancellationToken = default)
    {
        // web parity: the section's row test fires { id } with no title/message, so the request carries no body
        // and the structured result body is returned unconditionally (even on a non-2xx receiver status).
        var request = new ApiRequest(WebhookChannelsRegistration.WebhookTestOperation, PathParams: PathFor(id));
        var response = await _api.SendAsync<JsonElement?>(request, cancellationToken).ConfigureAwait(false);

        return response is { } value
            ? WebhookTestResult.FromJson(value)
            : new WebhookTestResult(false, 0, 0, null, null, false, null);
    }

    /// <inheritdoc />
    public async Task<string> PreviewSignatureAsync(string secret, string body, CancellationToken cancellationToken = default)
    {
        var payload = new JsonObject
        {
            ["secret"] = secret,
            ["body"] = body,
        };

        var request = new ApiRequest(WebhookChannelsRegistration.SignaturePreviewOperation, Body: payload);
        var response = await _api.SendAsync<JsonElement?>(request, cancellationToken).ConfigureAwait(false);

        if (response is { } value && value.ValueKind == JsonValueKind.Object)
        {
            return JsonScalars.ReadString(value, "signature") ?? string.Empty;
        }

        return string.Empty;
    }

    private static Dictionary<string, string> PathFor(long id) =>
        new(StringComparer.Ordinal)
        {
            [WebhookChannelsRegistration.ChannelIdParam] = id.ToString(CultureInfo.InvariantCulture),
        };

    // The cache-then-network "empty" predicate runs over the full GET /notifications array; the surface is empty
    // when it holds no kind=webhook rows (the web useWebhookChannels filter), even if other channel kinds exist.
    private static bool IsEmptyWebhooks(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Array)
        {
            return true;
        }

        foreach (var row in element.EnumerateArray())
        {
            if (row.ValueKind == JsonValueKind.Object &&
                string.Equals(JsonScalars.ReadString(row, "kind"), "webhook", StringComparison.OrdinalIgnoreCase))
            {
                return false;
            }
        }

        return true;
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/WebhookChannelsSection/WebhookChannelsSectionSource.kt
// The data port the WebhookChannelsSection feature view binds to (P1/S8 state-holder seam) — the native analogue
// of the web component's hook composition
// (web/src/api/hooks/useNotificationChannels.ts -> web/src/features/settings/components/WebhookChannelsSection.tsx).
// The view never performs HTTP itself; a shared adapter (the S8 stores or the S7 repositories) or a test fake
// drives this. Cache-then-network freshness is preserved end to end (ADR-013): every read emission's
// cached/stale/error flags flow through unchanged so the view-model can render the full state matrix.
//
// The six operations map one-to-one onto the web hooks the section uses: the filtered read is `useWebhookChannels`
// (the `useNotificationChannels` list filtered to `kind === 'webhook'`); [saveChannel]/[deleteChannel]/
// [toggleChannel] are the generic `useSaveChannel`/`useDeleteChannel`/`useToggleChannel` (they refresh the shared
// channel list, so the filtered read updates without cross-store plumbing); [testWebhookChannel] is the HMAC-aware
// `useTestWebhookChannel`; [previewWebhookSignature] is `useWebhookSignaturePreview`. Test + preview invalidate
// nothing (their result is rendered inline), exactly like the web mutations.
//
// `InvalidPackageDeclaration`/`filename`/`MatchingDeclarationName` are suppressed: the mandated surface directory
// (com/teslasync/feature-views/WebhookChannelsSection) cannot form a valid Kotlin package and the file hosts the
// seam plus its bindings, mirroring the sibling surfaces.
@file:Suppress("InvalidPackageDeclaration", "ktlint:standard:filename", "MatchingDeclarationName")

package io.teslasync.android.featureviews.webhookchannelssection

import io.teslasync.shared.core.data.repo.NotificationChannelsRepository
import io.teslasync.shared.core.data.repo.NotificationsRepository
import io.teslasync.shared.core.data.repo.Resource
import io.teslasync.shared.core.data.repo.filterWebhookChannels
import io.teslasync.shared.core.presentation.notificationchannels.NotificationChannel
import io.teslasync.shared.core.presentation.notificationchannels.NotificationChannelsStore
import io.teslasync.shared.core.presentation.notificationchannels.WebhookSignaturePreviewResult
import io.teslasync.shared.core.presentation.notificationchannels.WebhookTestResult
import io.teslasync.shared.core.presentation.notifications.NotificationChannelInput
import io.teslasync.shared.core.presentation.notifications.NotificationsStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

/**
 * The single seam the [WebhookChannelsSectionViewModel] depends on so it binds to an abstraction (real adapter ↔
 * test fake), never to a concrete store or the network. [webhookChannels] is the cache-then-network feed the web
 * `useWebhookChannels` serves (already filtered to webhook rows); [invalidate] is the web
 * `useInvalidateWebhookChannels` re-fetch trigger; the five mutations mirror the web `useSaveChannel` /
 * `useDeleteChannel` / `useToggleChannel` / `useTestWebhookChannel` / `useWebhookSignaturePreview` non-throwing
 * results. No HTTP touches the view.
 */
interface WebhookChannelsSectionSource {
    /** Stream the cache-then-network webhook-channel list (web `useWebhookChannels`, the `kind === 'webhook'` filter). */
    fun webhookChannels(): Flow<Resource<List<NotificationChannel.Webhook>>>

    /** Re-fetch the webhook-channel feed (web `useInvalidateWebhookChannels`); a no-op for a feed nobody observes. */
    fun invalidate()

    /** Create or update a channel (web `useSaveChannel`); refreshes the channel list on success. */
    suspend fun saveChannel(input: NotificationChannelInput): Result<NotificationChannel>

    /** Delete a channel (web `useDeleteChannel`); refreshes the channel list + stats on success. */
    suspend fun deleteChannel(id: Long): Result<Unit>

    /** Toggle a channel's enabled state (web `useToggleChannel`); refreshes the channel list + stats. */
    suspend fun toggleChannel(id: Long): Result<NotificationChannel>

    /**
     * Fire a structured webhook test through the HMAC-aware delivery path (web `useTestWebhookChannel`). Returns
     * the structured result even for a non-2xx receiver (the endpoint answers HTTP 200 in every delivery outcome);
     * invalidates nothing — the row renders the result inline.
     */
    suspend fun testWebhookChannel(
        id: Long,
        title: String?,
        message: String?,
    ): Result<WebhookTestResult>

    /**
     * Compute the `X-TeslaSync-Signature` for a `(secret, body)` pair (web `useWebhookSignaturePreview`). A pure
     * utility; invalidates nothing. An empty secret is rejected 400 server-side (callers guard first).
     */
    suspend fun previewWebhookSignature(
        secret: String,
        body: String,
    ): Result<WebhookSignaturePreviewResult>
}

/**
 * Binds the surface to the shared **S8** holders: the [NotificationChannelsStore] owns the filtered webhook read,
 * the HMAC test, the signature preview, and the invalidate trigger (web `useNotificationChannels.ts`), while the
 * generic [NotificationsStore] owns the channel CRUD (web `useSaveChannel`/`useDeleteChannel`/`useToggleChannel`).
 * CRUD routes through [NotificationsStore], which refreshes the shared `channels` feed the webhook read derives
 * from; the view-model additionally calls [invalidate] so the dedicated webhook feed re-fetches uniformly. No HTTP
 * touches the view — the stores (S7/S8) own it.
 */
fun webhookChannelsSectionSource(
    channelsStore: NotificationChannelsStore,
    notificationsStore: NotificationsStore,
): WebhookChannelsSectionSource =
    object : WebhookChannelsSectionSource {
        override fun webhookChannels(): Flow<Resource<List<NotificationChannel.Webhook>>> = channelsStore.webhookChannels()

        override fun invalidate() = channelsStore.invalidateWebhookChannels()

        override suspend fun saveChannel(input: NotificationChannelInput): Result<NotificationChannel> =
            notificationsStore.saveChannel(input)

        override suspend fun deleteChannel(id: Long): Result<Unit> = notificationsStore.deleteChannel(id)

        override suspend fun toggleChannel(id: Long): Result<NotificationChannel> = notificationsStore.toggleChannel(id)

        override suspend fun testWebhookChannel(
            id: Long,
            title: String?,
            message: String?,
        ): Result<WebhookTestResult> = channelsStore.testWebhookChannel(id, title, message)

        override suspend fun previewWebhookSignature(
            secret: String,
            body: String,
        ): Result<WebhookSignaturePreviewResult> = channelsStore.previewWebhookSignature(secret, body)
    }

/**
 * Binds the surface directly to the shared **S7** repositories. The [NotificationChannelsRepository] serves the
 * raw `GET /notifications` list (filtered to webhook rows here via [filterWebhookChannels], the web
 * `useWebhookChannels`) plus the HMAC test + signature preview; the generic [NotificationsRepository] serves the
 * channel CRUD. Each [webhookChannels] call starts a NEW cache-then-network collection, so the view-model's
 * refresh/retry trigger a genuine re-fetch (the web `refetch()` behaviour) and [invalidate] is a no-op — the
 * binding to use when a host does not share app-wide stores.
 */
fun webhookChannelsSectionSource(
    channelsRepository: NotificationChannelsRepository,
    notificationsRepository: NotificationsRepository,
): WebhookChannelsSectionSource =
    object : WebhookChannelsSectionSource {
        override fun webhookChannels(): Flow<Resource<List<NotificationChannel.Webhook>>> =
            channelsRepository.channels().map { it.mapData(::filterWebhookChannels) }

        override fun invalidate() = Unit

        override suspend fun saveChannel(input: NotificationChannelInput): Result<NotificationChannel> =
            notificationsRepository.saveChannel(input)

        override suspend fun deleteChannel(id: Long): Result<Unit> = notificationsRepository.deleteChannel(id)

        override suspend fun toggleChannel(id: Long): Result<NotificationChannel> = notificationsRepository.toggleChannel(id)

        override suspend fun testWebhookChannel(
            id: Long,
            title: String?,
            message: String?,
        ): Result<WebhookTestResult> = channelsRepository.testWebhookChannel(id, title, message)

        override suspend fun previewWebhookSignature(
            secret: String,
            body: String,
        ): Result<WebhookSignaturePreviewResult> = channelsRepository.previewWebhookSignature(secret, body)
    }

/** Maps a [Resource]'s `data` and `cached` slots through [f], preserving the freshness flags (ADR-013). */
private fun <A, B> Resource<A>.mapData(f: (A) -> B): Resource<B> =
    when (this) {
        is Resource.Loading -> Resource.Loading(cached?.let(f), fetchedAt, stale)
        is Resource.Success -> Resource.Success(f(data), fetchedAt, stale)
        is Resource.Error -> Resource.Error(cached?.let(f), fetchedAt, stale, error)
    }
